keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,534 @@
+    // Enabled indicates whether Job tracking uses this struct.
+    Enabled   bool
+    // Succeeded holds UIDs of succeeded Pods.
+    Succeeded []string
+    // Succeeded holds UIDs of failed Pods.
+    Failed    []string
+}
+```
+
+### Algorithm
+
+The following algorithm updates the status counters without relying on finished
+Pods to be present indefinitely. The algorithm assumes that the Job controller
+could be stopped at any point and executed again from the first step without
+losing information. Generally, all the steps happen in a single Job sync
+cycle.
+
+1. The Job controller calculates the number of succeeded Pods as the sum of:
+   - `.status.succeeded`,
+   - the size of `job.status.uncountedTerminatedPodUIDs.succeeded` and
+   - the number of finished Pods that are not in `job.status.uncountedTerminatedPodUIDs.succeeded`.
+   This number informs the creation of missing Pods to reach `.spec.completions`.
+   The controller creates Pods for a Job with the finalizer
+   `batch.kubernetes.io/job-completion`.
+2. The Job controller adds Pod UIDs to the `.status.uncountedTerminatedPodUIDs.succeeded`
+   and `.status.uncountedTerminatedPodUIDs.failed` lists if the Pod:
+    - has the `batch.kubernetes.io/job-completion` finalizer, and
+    - the Pod is on Succeeded or Failed phase, respectively.
+    The controller sends a status update.
+3. The Job controller removes the `batch.kubernetes.io/job-completion` finalizer
+   from all Pods on Succeeded or Failed phase that were added to the lists in
+   `.status.uncountedTerminatedPodUIDs` in the previous step. The controller keeps note
+   of the Pod updates that succedeed.
+4. The Job controller counts the Pods in the `.status.uncountedTerminatedPodUIDs` lists
+   that:
+   - had no finalizer at the beginning of the sync, or
+   - had the patch in step 3 succeed, or
+   - were removed from the system.
+   The counts increment the `.status.failed` and `.status.succeeded` and clears
+   counted Pods from `.status.uncountedPodsUIDs` lists. The controller sends a
+   status update.
+
+Steps 2 to 4 might deal with a potentially big number of Pods. Thus, status
+updates can potentially stress the kube-apiserver. For this reason, the Job
+controller repeats steps 2 to 4, capping the number of Pods each time, until
+the controller processes all the Pods. The number of Pods is caped by:
+- time: in each iteration, the job controller removes all the Pods' finalizers
+  it can in 1 minute. This allows to throttle the number of status updates.
+- count: Preventing big writes to etcd.
+
+### User deleted Pods
+   
+In the case where a user or another controller removes a Pod, which sets a
+deletion timestamp, the Job controller removes the finalizer from the Pod,
+<<[UNRESOLVED: should we count terminating pods?]>>
+independently of it having finished or not.
+This allows for deleted Pods to not count towards the failed counter, so it
+doesn't change the current behavior.
```

Comment:
The current controller can't count pods that don't exist :(

Reply:
Ack